lib/puppet/http/client.rb

```diff
@@ -404,6 +404,13 @@ def execute_streaming(request, options: {}, &block)
 
           done = true
         end
+      ensure
+        # Connections cannot be closed inside a block passed to http.request,
+        # so close it after the function has been completed.
+        if close_connection_after_request
+          Puppet.debug("Closing connection for #{Puppet::HTTP::Site.from_uri(request.uri)}")
+          http.finish
+        end
```

Comment:
Yeah moving `http.finish` outside makes sense. I'm a little concerned about sleeping **before** closing the connection. It's not a specific concern just a general unease about half-open connections. For example, if puppetserver responds with 503 it will tell the agent to sleep for [1800 seconds by default](https://github.com/puppetlabs/puppetserver/blob/2a6cd29df1979313490a099d1cef4036450baef7/src/clj/puppetlabs/services/request_handler/request_handler_service.clj#L20). While the agent is sleeping, a load-balancer in front of puppetservers could timeout the idle connection after 30 seconds - 2 minutes. So then we have a half-open situation where the client may think the connection is open, but the server closed it, and the client won't know until it tries to flush buffered data, and fails. 

To minimize risk of breaking something in 7.x, maybe move the entire "finish-and-sleep" code block outside of the `http.request` method, something like?

```ruby
while !done do
  interval = nil
  connect(request.uri, options: options) do |http|
    ...
    http.request(request) do |nethttp|
      response = Puppet::HTTP::ResponseNetHTTP.new(request.uri, nethttp)
      begin
        ...
        interval = @retry_after_handler.retry_after_interval(request, response, retries)
        ...      
      ensure
        response.drain
      end

      done = true unless interval
    end
  ensure
    if interval
      if http.started?
        Puppet.debug("Closing connection for #{Puppet::HTTP::Site.from_uri(request.uri)}")
        http.finish
      end
      Puppet.warning(_("Sleeping for %{interval} seconds before retrying the request") % { interval: interval })
      ::Kernel.sleep(interval)
    end
  end
end
```

Reply:
Yeah, good catch about the sleeping before the close; I've refactored that into `block` let me know if you think that makes sense.

Reply:
Ah yeah, nice!